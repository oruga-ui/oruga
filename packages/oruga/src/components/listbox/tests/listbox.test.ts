import { afterEach, describe, expect, test } from "vitest";
import { enableAutoUnmount, mount } from "@vue/test-utils";

import OListbox from "@/components/listbox/Listbox.vue";
import type { OptionsProp } from "@/composables";

describe("OListbox tests", () => {
    enableAutoUnmount(afterEach);

    const options: OptionsProp = [
        { label: "New York", value: "NY" },
        { label: "Rome", value: "RM" },
        { label: "London", value: "LDN" },
        { label: "Istanbul", value: "IST" },
        { label: "Paris", value: "PRS" },
    ];

    test("render correctly", () => {
        const wrapper = mount(OListbox, {
            props: { options },
        });
        expect(!!wrapper.vm).toBeTruthy();
        expect(wrapper.exists()).toBeTruthy();
        expect(wrapper.attributes("data-oruga")).toBe("listbox");
        expect(wrapper.html()).toMatchSnapshot();
    });

    test("should exist", () => {
        const wrapper = mount(OListbox, {
            props: { options },
        });
        expect(wrapper.find(".p-listbox.p-component").exists()).toBe(true);
        expect(wrapper.findAll("li.p-listbox-option").length).toBe(5);
        expect(
            wrapper.findAll("li.p-listbox-option")[0].attributes()[
                "aria-label"
            ],
        ).toBe("New York");
    });

    test("should select a list item", async () => {
        const wrapper = mount(OListbox, {
            props: { options },
        });
        await wrapper.vm.onOptionSelect({}, wrapper.vm.options[0]);

        expect(wrapper.emitted()["update:modelValue"][0]).toEqual([
            wrapper.vm.options[0],
        ]);

        await wrapper.setProps({ modelValue: wrapper.vm.options[0] });

        expect(wrapper.findAll("li.p-listbox-option")[0].classes()).toContain(
            "p-listbox-option-selected",
        );
    });

    describe("filter", () => {
        test("should have correct custom icon", async () => {
            const wrapper = mount(OListbox, {
                props: { options },
            });
            await wrapper.setProps({
                filter: true,
                filterIcon: "pi pi-discord",
            });

            const icon = wrapper.find(
                '.p-inputicon [data-pc-section="filtericon"]',
            );

            expect(icon.classes()).toContain("pi-discord");
        });

        test("should correctly filter", async () => {
            const wrapper = mount(OListbox, {
                props: { options },
            });
            await wrapper.setProps({
                filter: true,
            });

            const filterInput = wrapper.find("input.p-listbox-filter");

            expect(filterInput.exists()).toBe(true);

            await filterInput.setValue("is");

            const options = wrapper.findAll(".p-listbox-option");

            expect(options.length).toBe(2);
            expect(options[0].text()).toBe("Istanbul");
        });

        test("should correctly filter groups", async () => {
            const wrapper = mount(OListbox, {
                props: { options },
            });
            await wrapper.setProps({
                filter: true,
                optionGroupLabel: "label",
                optionLabel: "label",
                optionGroupChildren: "items",
                options: [
                    {
                        label: "Germany",
                        code: "DE",
                        items: [
                            { label: "Berlin", value: "Berlin" },
                            { label: "Frankfurt", value: "Frankfurt" },
                            { label: "Hamburg", value: "Hamburg" },
                            { label: "Munich", value: "Munich" },
                        ],
                    },
                    {
                        label: "USA",
                        code: "US",
                        items: [
                            { label: "Chicago", value: "Chicago" },
                            { label: "Los Angeles", value: "Los Angeles" },
                            { label: "New York", value: "New York" },
                            { label: "San Francisco", value: "San Francisco" },
                        ],
                    },
                ],
            });

            const filterInput = wrapper.find("input.p-listbox-filter");

            expect(filterInput.exists()).toBe(true);

            await filterInput.setValue("ch");

            const optionGroups = wrapper.findAll(".p-listbox-option-group");
            const options = wrapper.findAll(".p-listbox-option");

            expect(optionGroups.length).toBe(2);
            expect(optionGroups[0].text()).toBe("Germany");
            expect(optionGroups[1].text()).toBe("USA");

            expect(options.length).toBe(2);
            expect(options[0].text()).toBe("Munich");
            expect(options[1].text()).toBe("Chicago");
        });
    });
});
