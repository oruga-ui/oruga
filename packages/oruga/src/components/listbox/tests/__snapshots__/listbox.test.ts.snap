// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OListbox tests > render correctly with items 1`] = `
"<div data-oruga="listbox" class="o-listbox o-listbox--selectable">
  <!--v-if-->
  <!--v-if-->
  <div class="o-listbox__container" style="max-height: 225px;">
    <ul id="v-0_list" role="listbox" tabindex="0" class="o-listbox__list" aria-disabled="false">
      <transition-group-stub name="fade" appear="false" persisted="false" css="true">
        <!--
                        @slot Place items here
                        @binding {number} focusedIndex - index of the focused element
                    -->
        <li id="v-0-1" data-oruga="listbox-item" data-id="listbox-1" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->New York
        </li>
        <li id="v-0-2" data-oruga="listbox-item" data-id="listbox-2" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->Rome
        </li>
        <li id="v-0-3" data-oruga="listbox-item" data-id="listbox-3" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->London
        </li>
        <li id="v-0-4" data-oruga="listbox-item" data-id="listbox-4" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->Istanbul
        </li>
        <li id="v-0-5" data-oruga="listbox-item" data-id="listbox-5" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->Paris
        </li>
      </transition-group-stub>
    </ul>
  </div>
  <transition-stub name="fade" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</div>"
`;

exports[`OListbox tests > render correctly with options 1`] = `
"<div data-oruga="listbox" class="o-listbox o-listbox--selectable">
  <!--v-if-->
  <!--v-if-->
  <div class="o-listbox__container" style="max-height: 225px;">
    <ul id="v-0_list" role="listbox" tabindex="0" class="o-listbox__list" aria-disabled="false">
      <transition-group-stub name="fade" appear="false" persisted="false" css="true">
        <!--
                        @slot Place items here
                        @binding {number} focusedIndex - index of the focused element
                    -->
        <!--v-if-->
        <li id="v-0-1" data-oruga="listbox-item" data-id="listbox-1" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false" aria-label="New York" aria-setsize="5" aria-posinset="1">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->
          <!--
                                    @slot Override the label, default is label prop
                                    @binding {object} option - option item
                                    @binding {number} index - option index
                                    @binding {boolean} selected - option is selected
                                    @binding {boolean} disabled - option is disabled
                                --><span>New York</span>
        </li>
        <li id="v-0-2" data-oruga="listbox-item" data-id="listbox-2" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false" aria-label="Rome" aria-setsize="5" aria-posinset="2">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->
          <!--
                                    @slot Override the label, default is label prop
                                    @binding {object} option - option item
                                    @binding {number} index - option index
                                    @binding {boolean} selected - option is selected
                                    @binding {boolean} disabled - option is disabled
                                --><span>Rome</span>
        </li>
        <li id="v-0-3" data-oruga="listbox-item" data-id="listbox-3" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false" aria-label="London" aria-setsize="5" aria-posinset="3">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->
          <!--
                                    @slot Override the label, default is label prop
                                    @binding {object} option - option item
                                    @binding {number} index - option index
                                    @binding {boolean} selected - option is selected
                                    @binding {boolean} disabled - option is disabled
                                --><span>London</span>
        </li>
        <li id="v-0-4" data-oruga="listbox-item" data-id="listbox-4" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false" aria-label="Istanbul" aria-setsize="5" aria-posinset="4">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->
          <!--
                                    @slot Override the label, default is label prop
                                    @binding {object} option - option item
                                    @binding {number} index - option index
                                    @binding {boolean} selected - option is selected
                                    @binding {boolean} disabled - option is disabled
                                --><span>Istanbul</span>
        </li>
        <li id="v-0-5" data-oruga="listbox-item" data-id="listbox-5" class="o-listbox__item" role="option" tabindex="-1" aria-selected="false" aria-hidden="false" aria-disabled="false" aria-label="Paris" aria-setsize="5" aria-posinset="5">
          <!--
            @slot Override the label, default is label prop
            @binding {boolean} selected - item is selected
            @binding {boolean} disabled - item is disabled
        -->
          <!--
                                    @slot Override the label, default is label prop
                                    @binding {object} option - option item
                                    @binding {number} index - option index
                                    @binding {boolean} selected - option is selected
                                    @binding {boolean} disabled - option is disabled
                                --><span>Paris</span>
        </li>
      </transition-group-stub>
    </ul>
  </div>
  <transition-stub name="fade" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</div>"
`;
